### PR TITLE
Scan commit metadata

### LINKED ISSUE #2683: [Request] Scan the Commit Metadata (email, name and commit message) for secrets
Hi Team,

## Description
Had read a blog were the author mentioned that user's might have secrets set in their git commit configurations. So I started logging the commit metadata for public repositories and found **two** instance of credentials being configured in the Commit email address. What looks like its occurs mainly due to mistake made while copy pasting the commands during git configuration.

One of the findings that I found in the wild
image

To replicate the same, I have set the hugginface API key in the email field which doesn't gets detected by the trufflehog https://github.com/bugbaba/testwa54321/commit/6819bdd07fba20de2a1249bc8bf5341530b53d4a.patch

image

### Preferred Solution
While parsing the git commits saved the commit metadata to a file which is later scanned for secrets.

--
Regards,
@bugbaba

### PR DESCRIPTION
**Pull request by rgmz**
### Description:
This fixes #2683. It scans the commit author, committer (which is typically `GitHub <noreply@github.com>` for GitHub, but can be different), and message.

```
commit 8104611d6e2e3993c104974ea4c507faced7e847
Author:     Richard Gomez <32133502+rgmz@users.noreply.github.com>
AuthorDate: Tue Feb 6 10:13:53 2024 -0500
Commit:     GitHub <noreply@github.com>
CommitDate: Tue Feb 6 10:13:53 2024 -0500

    fix: case-insensitive ext check (#2383)
```

Admittedly, this implementation is a bit of guess-work and may not be the best way.

### Checklist:
* [x] Tests passing (`make test-community`)?
* [ ] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

### CODE REVIEW

File: pkg/gitparse/gitparse.go [hunk 1]
```diff
@@ -209,7 +214,14 @@ func NewParser(options ...Option) *Parser {
...
 // The Diff chan will return diffs in the order they are parsed from the log.
 func (c *Parser) RepoPath(ctx context.Context, source string, head string, abbreviatedLog bool, excludedGlobs []string, isBare bool) (chan *Diff, error) {
-	args := []string{"-C", source, "log", "-p", "--full-history", "--date=format:%a %b %d %H:%M:%S %Y %z"}
+	args := []string{
+		"-C", source,
+		"log",
+		"--patch", // https://git-scm.com/docs/git-log#Documentation/git-log.txt---patch
+		"--full-history",
+		"--date=format:%a %b %d %H:%M:%S %Y %z",
+		"--pretty=fuller", // https://git-scm.com/docs/git-log#_pretty_formats
```
Comment (rosecodym) [inline pkg/gitparse/gitparse.go:223]: Are we worried that the pre-canned pretty formats might change in the future?

File: pkg/gitparse/gitparse.go [hunk 2]
```diff
@@ -577,20 +594,42 @@ func isAuthorLine(isStaged bool, latestState ParseState, line []byte) bool {
...
+	if isStaged || latestState != CommitterLine {
+		return false
+	}
+	if len(line) > 10 && bytes.Equal(line[:11], []byte("CommitDate:")) {
 		return true
 	}
 	return false
 }
 
 // Line directly after Date with only a newline.
```
Comment (rosecodym) [inline pkg/gitparse/gitparse.go:630]: update? (`Date` -> `CommitDate`)

File: pkg/gitparse/gitparse.go [hunk 3]
```diff
@@ -373,16 +385,21 @@ func (c *Parser) FromReader(ctx context.Context, stdOut io.Reader, diffChan chan
...
 			if err != nil {
-				ctx.Logger().V(2).Info("Could not parse date from git stream.", "error", err)
+				ctx.Logger().Error(err, "failed to parse commit date", "commit", currentCommit.Hash)
 			}
 			currentCommit.Date = date
+		case isCommitterLine(isStaged, latestState, line):
+			latestState = CommitterLine
+			currentCommit.Committer = strings.TrimSpace(string(line[8:]))
+		case isCommitterDateLine(isStaged, latestState, line):
+			latestState = CommitterDateLine
```
Comment (rosecodym) [inline pkg/gitparse/gitparse.go:401]: Is there any particular reason you called the state "Commit**ter**DateLine" when the Git output is "CommitDate"?

File: pkg/sources/git/git.go [hunk 4]
```diff
@@ -562,29 +562,55 @@ func (s *Git) ScanCommits(ctx context.Context, repo *git.Repository, path string
...
 			atomic.AddUint64(&s.metrics.commitsScanned, 1)
 			logger.V(5).Info("scanning commit", "commit", fullHash)
-		}
 
-		if !scanOptions.Filter.Pass(diff.PathB) {
-			continue
+			// Scan the commit metadata.
+			// See https://github.com/trufflesecurity/trufflehog/issues/2683
+			var (
+				metadata = s.sourceMetadataFunc("", email, fullHash, when, remoteURL, 0)
```
Comment (rosecodym) [inline pkg/sources/git/git.go:584]: Can we use a magic constant filename to signal to users that the chunk came from commit metadata? I'm a little worried that a missing filename might confuse people.

File: pkg/sources/git/git.go [hunk 5]
```diff
@@ -562,29 +562,55 @@ func (s *Git) ScanCommits(ctx context.Context, repo *git.Repository, path string
...
 		fileName := diff.PathB
 		if fileName == "" {
 			continue
 		}
-		email := diff.Commit.Author
-		when := diff.Commit.Date.UTC().Format("2006-01-02 15:04:05 -0700")
+
+		if !scanOptions.Filter.Pass(fileName) {
+			continue
+		}
```
Comment (rosecodym) [inline pkg/sources/git/git.go:613]: What's this for?

**Review by rosecodym — COMMENTED** [review]
This is cool! I never would have guessed that secrets could leak out in the author email.

File: pkg/sources/git/git.go (hunk 5 above)
Comment (rgmz) [reply pkg/sources/git/git.go:613]: I'm not sure. This is an existing piece of code that got moved around (albeit, with the minor change of reusing the  declared `fileName` variable.)

File: pkg/gitparse/gitparse.go (hunk 3 above)
Comment (rgmz) [reply pkg/gitparse/gitparse.go:401]: It's to disambiguate the committ**er** information:
```
Commit: Dustin Decker <dustin@trufflesec.com>
CommitDate: Tue Feb 6 10:13:53 2024 -0500
```

With the commit information:
```
commit 4218c39d99b5f30153f62471c1be1c1596f0a4d4
```

I don't know why they have `Author` and `Commit` instead of `Committer`, when committer is used explicitly in the docs. Perhaps they wanted to keep the line lengths the same?

File: pkg/gitparse/gitparse.go (hunk 1 above)
Comment (rgmz) [reply pkg/gitparse/gitparse.go:223]: I'm not worried. The formats seem to be consistent. 
https://web.archive.org/web/20140208065758/https://git-scm.com/docs/git-log

File: pkg/sources/git/git.go (hunk 4 above)
Comment (rgmz) [reply pkg/sources/git/git.go:584]: We could do something obvious like `COMMIT`. I will mention that empty filename is currently how this is handled elsewhere, e.g., comments

File: pkg/gitparse/gitparse.go [hunk 6]
```diff
@@ -22,7 +22,7 @@ import (
 
 const (
 	// defaultDateFormat is the standard date format for git.
-	defaultDateFormat = "Mon Jan 02 15:04:05 2006 -0700"
+	defaultDateFormat = "Mon Jan 2 15:04:05 2006 -0700"
```
Comment (rgmz) [inline pkg/gitparse/gitparse.go:25]: I believe this is a bug as `git log` appears to use `d` and not `dd` formatting. It's unclear if or how frequently this failure occurred, as the logging in `isDateLine`/`isAuthorDateLine` was level 2.

File: pkg/gitparse/gitparse.go [hunk 7]
```diff
@@ -622,15 +681,51 @@ func isMessageEndLine(isStaged bool, latestState ParseState, line []byte) bool {
...
 	if !(latestState == MessageStartLine || // Empty commit messages can go from MessageStart->Diff
 		latestState == MessageEndLine ||
+		latestState == NotesEndLine ||
 		latestState == BinaryFileLine ||
+		latestState == ModeLine ||
 		latestState == IndexLine ||
 		latestState == HunkContentLine ||
 		latestState == ParseFailure) {
-		if latestState == Initial && !isStaged {
+		if !(isStaged && latestState == Initial) {
```
Comment (rgmz) [inline pkg/gitparse/gitparse.go:728]: This is an unrelated bug that caused `isDiffLine` to return true for any `latestState`, not just the ones mentioned in the list.

**Review by rgmz — COMMENTED** [review]
I pushed some changes to handle [Git Notes](https://tylercipriani.com/blog/2022/11/19/git-notes-gits-coolest-most-unloved-feature/) as well. However, these may not be pulled by default and would need to be added as an additional ref to checkout in #1918.

File: pkg/sources/git/git.go (hunk 4 above)
Comment (rosecodym) [reply pkg/sources/git/git.go:584]: Ah, if this isn't a new approach I'm less concerned about it, but it still seems like if it's low-hanging fruit we might as well grab it while we're here. I'll leave the decision on fruit height up to you.

File: pkg/gitparse/gitparse.go (hunk 7 above)
Comment (rosecodym) [reply pkg/gitparse/gitparse.go:728]: Was there any real impact of that? Or it would it only flare up if the state machine got into a bad state in some other way?

File: pkg/gitparse/gitparse.go (hunk 3 above)
Comment (rosecodym) [reply pkg/gitparse/gitparse.go:410]: that's...very strange. but ok!

**Review by rosecodym — APPROVED** [review]
cool stuff!

File: pkg/gitparse/gitparse.go (hunk 3 above)
Comment (rgmz) [reply pkg/gitparse/gitparse.go:410]: ¯\\_(ツ)_/¯

**Comment by rgmz** [comment]
Linting issues seem to be carried over from #2643.

https://github.com/trufflesecurity/trufflehog/actions/runs/8802614804

**Comment by rgmz** [comment]
@rosecodym Any idea what might be causing the [`test` failure](https://github.com/trufflesecurity/trufflehog/actions/runs/8834183763/job/24255263367)?
```
=== Failed
=== FAIL: pkg/sources/git TestSource_Chunks_Integration/remote_repo,_unauthenticated (0.23s)
2024-04-25T14:15:48Z	error	context	error chunking	{"error": "context canceled"}
2024-04-25T14:15:48Z	info-0	context	error scanning repository	{"repo": "../../../", "error": "context canceled"}
2024-04-25T14:15:48Z	error	context	error chunking	{"error": "context canceled"}
2024-04-25T14:15:48Z	info-0	context	error scanning repository	{"repo": "https://github.com/dustin-decker/secretsandstuff.git", "error": "context canceled"}
2024-04-25T14:15:48Z	error	context	error chunking	{"error": "context canceled"}
2024-04-25T14:15:48Z	info-0	context	error scanning repository	{"repo": "https://github.com/dustin-decker/secretsandstuff.git", "error": "context canceled"}
2024-04-25T14:15:48Z	error	context	error chunking	{"error": "context canceled"}
2024-04-25T14:15:48Z	info-0	context	error scanning repository	{"repo": "https://github.com/dustin-decker/secretsandstuff.git", "error": "context canceled"}
    git_test.go:254: A chunk exists that was not expected with key "70001020fab32b1fcf2f1f0e5c66424eae649826-"
    git_test.go:254: A chunk exists that was not expected with key "a6f8aa55736d4a85be31a0048a4607396898647a-"
    git_test.go:254: A chunk exists that was not expected with key "73ab4713057944753f1bdeb80e757380e64c6b5b-"
    git_test.go:254: A chunk exists that was not expected with key "2f251b8c1e72135a375b659951097ec7749d4af9-"
    git_test.go:254: A chunk exists that was not expected with key "84e9c75e388ae3e866e121087ea2dd45a71068f2-"
    git_test.go:254: A chunk exists that was not expected with key "e6c8bbabd8796ea3cd85bfc2e55b27e0a491747f-"
    git_test.go:254: A chunk exists that was not expected with key "735b52b0eb40610002bb1088e902bd61824eb305-"
    git_test.go:254: A chunk exists that was not expected with key "ce62d79908803153ef6e145e042d3e80488ef747-"
    git_test.go:254: A chunk exists that was not expected with key "27fbead3bf883cdb7de9d7825ed401f28f9398f1-"
    git_test.go:254: A chunk exists that was not expected with key "8afb0ecd4998b1179e428db5ebbcdc8221214432-"
    git_test.go:254: A chunk exists that was not expected with key "8fe6f04ef1839e3fc54b5147e3d0e0b7ab971bd5-"

=== FAIL: pkg/sources/git TestSource_Chunks_Integration/remote_repo,_limited (0.19s)
    git_test.go:254: A chunk exists that was not expected with key "70001020fab32b1fcf2f1f0e5c66424eae649826-"

=== FAIL: pkg/sources/git TestSource_Chunks_Integration/remote_repo,_main_ahead_of_branch (0.20s)
    git_test.go:254: A chunk exists that was not expected with key "547865c6cc0da46622306902b1b66f7e25dd0412-"
```

**Comment by rosecodym** [comment]
@rgmz should we revert this until we can figure out what's going on?

**Comment by rgmz** [comment]
Perhaps. The specific error seems to be caused by the test logic (commit + - + file); it's unclear whether that's indicative of an actual issue.

https://github.com/trufflesecurity/trufflehog/blob/81a9c813a17490aedbead3323493d687af14eea9/pkg/sources/git/git_test.go#L249-L251

**Comment by rosecodym** [comment]
Ok, I'm going to revert. Thanks for catching this so early!
